### DISCUSSION TITLE
MongoDB Transactions

### DISCUSSION THREAD

**Discussion by msavin** [meteor/meteor#11737] [Feature Requests]
Hey everyone,

It's still hard for me to believe that Meteor doesn't support MongoDB transactions. It was a really big feature when launched, and can potentially broaden the appeal of Meteor. 

Looking from their API, it's a pretty simple set of steps to use:
1. start the transaction
2. list all the mongo operations you wish to perform
3. submit the transaction and wait for the response

Good reference with examples of how to use it: https://forums.meteor.com/t/solved-transactions-with-mongodb-meteor-methods/48677

For a Meteor implementation, we need to have the Fibers support working and maybe add a bit of sugar to it? At minimum, we shouldn't have to rely on internal APIs to have it working.

A few ideas to start the conversation...

```js
Meteor.methods({
	"example1_purchaseItem": function (purchaseData) {
		const { invoiceId } = Mongo.Transaction(() => {
			const invoiceId = Invoice.insert({...purchaseData});
			const changeQuantity = Items.update(purchaseData.itemId, { ... })
			return { invoiceId, changeQuantity }
		}, errorCallback)

		return invoiceId;
	},
	"example2_purchaseItem": function (purchaseData) {
		// Everything in between transaction start and resolve shou
		Mongo.transaction.start("transaction1");

		const invoiceId = Invoice.insert({...purchaseData});
		const changeQuantity = Items.update(purchaseData.itemId, { ... })

		Mongo.transaction.resolve("transaction1");

		// this is not a part of the transaction
		Meteor.users.update({ ... })
		
		return invoiceId;
	}
})
```

I'm personally a fan of the example1, as anyone can pick up its use just seeing from a code sample, as well as easily migrate their existing code to support it.

-----

I also have a specific use case with my jobs queue package, [Steve Jobs](https://github.com/msavin/SteveJobs), where I think this can really shine. The job state is stored in MongoDB, and the idea is, to perform all the DB operations that the job requires, and update the job state document, in one shot. For example:

- Stripe sends a webhook message that a charge was successful
- Meteor server registers a job to process this information
- In one DB operation, the  job can then: 
     - update the payment status
     - mark the order as ready for delivery
     -  update the job document to say it was resolved successfully

**Comment by make-github-pseudonymous-again** [comment]
Using [the async/await solution suggested in the forums](https://forums.meteor.com/t/solved-transactions-with-mongodb-meteor-methods/48677), one can get:

```ts
Meteor.methods({
  'some.method.name'(itemId: string, smurfId: string) {
    return executeTransaction(async () => {
      const smurf = await Smurfs.rawCollection().findOne(
        {_id: smurfId},
        {session},
      );
      if (!smurf) {
        throw new Meteor.Error('not-found');
      }

      const meta = await Items.rawCollection().updateOne(
        {_id: itemId},
        {
          $pull: {'meta.y': smurfId},
        },
        {session},
      );

      ...

    });
  },
});
```

Reactivity still works, although maybe lag is larger, Meteor is able to pickup changes from Mongo's oplog (correct?). One can even throw in specific transaction options or session options if needed. What's annoying is that one has to pass the `session` around (forgetting one will just make the thing not atomic silently?)  and specifying `rawCollection()` all the time. One could make wrapper objects for Meteor collections so that ALL database interactions (`insert`, `update`, `find`, ...) magically forward it to `rawCollection` with the `session` but I did not go all the way there yet. Not sure what's a not too wasteful way to do it: we need to wrap ALL database interaction for ALL involved collections but most of my transactions will only use two or three verbs... Maybe a lazy getter. I do not know.

**PS**: I personally prefer try/catch and surf the async vibe:
```ts
"example3_purchaseItem": async function (purchaseData) {
  try {
    const { invoiceId } = await Mongo.Transaction(async () => {
      const invoiceId = await Invoice.insert({...purchaseData});
      const changeQuantity = await Items.update(purchaseData.itemId, { ... })
      return { invoiceId, changeQuantity }
    });
    return invoiceId;
  } catch (error: unknown) {
    ...
  }
},
```

**Reply by msavin** [reply]
Yep - if the lag is longer, its simply because of the transaction performance. Meteor should immediately pick up the changes regardless of how they were made.

**Reply by Torgen** [reply]
Anytime you update a collection outside of meteor's wrapper inside a method, call `Meteor.refresh({collection, id})`. It will make the method wait for the update to that document to be delivered to the client before returning.

**Reply by davidsavoie1** [reply]
@Torgen, where could I get some more information about `Meteor.refresh`? It seems undocumented...

Moreover, it seems I would need to provide a document id. If I use a transaction to update many documents at once (with `updateMany` for example, or multiple documents in different collections), would I then have to retrieve each document id and call `Meteor.refresh` for each?

Thanks for your feedback!

**Reply by Torgen** [reply]
It occurred to me that this might refer to a problem that is different from the one you have. The mechanism the subscription/method infrastructure uses to ensure that the effects of a method are published to the caller of that method on any relevant subscriptions before the method returns is `Meteor.refresh`. This is most relevant if your method has simulation, because without it the simulation will change the client side data, then the method will be called, the results will be returned without the side effects being published, the simulated changes will be rolled back, then later the effects will be received by oplog/polling and will become visible on the client side again. If you're using `rawCollection` and not separately implementing a client-side version, then the simulation roll back/roll forward flicker won't happen. You may still want to do it in that case if it's useful for the subscriptions to be updated before the method returns, but that doesn't guarantee a latency improvement.

Despite not having an underscore name, it seems like `Meteor.refresh` is very much internal and undocumented; from what I can tell, it takes a dict which is the selectors for updated documents, except that `_id` needs to be changed to `id` and you need a `collection` entry matching the collection name. It seems like you would need to collect all your selectors from your transaction and refresh them at the end. This is another reason why built-in transaction support would be better than a patched-on solution like the mongo-transactions package, since the collection methods could do that for you and refresh only after the transaction is committed.

**Reply by make-github-pseudonymous-again** [reply]
> @Torgen This is most relevant if your method has simulation, because without it the simulation will change the client side data, then the method will be called, the results will be returned without the side effects being published, the simulated changes will be rolled back, then later the effects will be received by oplog/polling and will become visible on the client side again.

This is clear, thanks a lot for the explanation!

> @Torgen  If you're using rawCollection and not separately implementing a client-side version, then the simulation roll back/roll forward flicker won't happen.

Here I guess you mean not executing anything on the client, so that no changes are simulated.

**Reply by make-github-pseudonymous-again** [reply]
> @Torgen Despite not having an underscore name, it seems like Meteor.refresh is very much internal and undocumented; from what I can tell, it takes a dict which is the selectors for updated documents, except that _id needs to be changed to id and you need a collection entry matching the collection name. It seems like you would need to collect all your selectors from your transaction and refresh them at the end.

This should be doable. I thought one had to list all the affected ids which would be hard. I see how this handles published data, but how does this non-published data that needs to be published due to changes (think pagination, newly created data, etc.).

> @Torgen This is another reason why built-in transaction support would be better than a patched-on solution like the mongo-transactions package, since the collection methods could do that for you and refresh only after the transaction is committed.

Agreed.

**Comment by diavrank** [comment]
Also consider to add transactions support for **Accounts methods** like createUser, sendEnrollmentEmail, sendVerificationEmail, etc.

**Reply by diavrank** [reply]
and if possible, add transactions for Astronomy package

**Reply by StorytellerCZ** [reply]
Astronomy package is a community package, for anything related to it you will need to reach out to Astronomy maintainers: https://github.com/jagi/meteor-astronomy/

**Reply by diavrank** [reply]
It seems to be that Lukasz no longer want to add new updates like this. So, I was wondering if Meteor Community can fork his repository to add this feature. More info [here](https://github.com/jagi/meteor-astronomy/pull/707).

**Reply by StorytellerCZ** [reply]
For that please start an issue here: https://github.com/Meteor-Community-Packages/organization/issues/new/choose

**Comment by radekmie** [comment]
I just want to point out one thing: handling of transaction failures (either transient or "actual" errors) is an important part of the business logic (you not always want to retry and sometimes you know it's safe to retry until success). "Hiding" it in the `Mongo.Transaction` or any other API would require an option to configure that.

That's why I'd rather have the access to the underlying API with an explicit creation of the `session` and passing it around. Whether or not we'll use Fibers or any other implicit storage (Fibers shouldn't be used since https://github.com/meteor/meteor/discussions/11505) is an implementation detail and can be handled in the user space (i.e., through a community maintained package).

Also, since https://github.com/meteor/meteor/pull/11605 is already in progress, that may be much easier to expose this functionality. The only problem is to make sure it'll play well with the existing code, like server-generated IDs and schema validation (collection packages in general).

**Reply by make-github-pseudonymous-again** [reply]
Do I understand correctly that the goal of a Meteor wrap around MongoDB transactions would be to allow transactions to be simulated on the client without writing twice the amount of code (isomorphism)? In that case, it seems natural that not every detail of MongoDB's implementation of transactions would transpire.

**Reply by radekmie** [reply]
It's possible to do the same without extra magic. If the `Mongo` object would provide an API to create the transaction (on the server it'd use the real one, on the client a mock or even a placeholder) and accept it in "correct places", it'd be enough. Also, the client-side version could actually verify, that you're using a `session` object created with this API (e.g., by using a tagged object).

On the other hand, I'm not really sure whether I like this idea. It may suggest, that the transactions will happen also on the client-side, and that's far from true.

**Reply by red-meadow** [reply]
One thing related to transaction failures is handling operations callbacks (`collection.insertOne(doc, callback)`). It seems reasonable to me that inside a transaction all callbacks should be called with either result (if the transaction was successful) or error argument. So they should be deferred until the transaction is ended. But last time I checked (2020), MongoDB Node.JS Driver was implementing just the opposite: callbacks were fired immediately, regardless of the transaction success. So it was easy to get false positive calls before a failed operation.
I know, the consistency is important, but maybe Meteor could do better in this regard?

**Reply by radekmie** [reply]
> I know, the consistency is important, but maybe Meteor could do better in this regard?

Well, I'd rather have it compliant with MongoDB instead of doing anything "better".

**Reply by bhunjadi** [reply]
Yeah, I'm not sure I get the whole point of deferred callbacks here and why would this be better in any way. I certainly wouldn't want the system to behave like that.

I wouldn't call these callbacks "false positive" because on atomic level they were legitimate; doc was inserted, albeit in the transaction that would eventually fail.

**Comment by msavin** [comment]
Maybe this discussion needs to be expanded into modernizing Meteor <-> MongoDB driver integration. I know there could be some breaking changes once Fibers are dropped. Maybe this can all make for a proper Meteor 3.0 release

**Comment by bhunjadi** [comment]
I wrote this [package](https://github.com/bhunjadi/meteor-mongo-transactions) for our app and it is working as expected for more than a year now. I also mentioned this in the thread you linked.

It's basically the same as you proposed in your first example with `Mongo.Transaction`.

Example:
```
function insertInvoice(invoice, lineItems) {
    runInTransaction(() => {
        Invoice.insert(invoice);
        LineItems.insert(lineItems);

        const payment = createPaymentDoc(); 
        Payments.insert(payment);
    });
}
```

**Reply by msavin** [reply]
Nice! I was just about to build something like this. It would be great to see the POC terminology dropped and some plain JS documentation.

**Reply by bhunjadi** [reply]
Ok, I might just do something about docs, but there isn't much to it besides `runInTransaction` function I already documented.

Also, there were some [cases](https://github.com/bhunjadi/meteor-mongo-transactions/blob/cec733a3d486e024b6bb5f659df60eba63ac0878/tests/server/server.tests.ts#L214) which I didn't manage to resolve within the package itself. Issue is with Mongo async callbacks and whether they are executed within the transaction.

Examples in tests are artificial since I don't think anybody writes code like that in Meteor. One real-world example is [denormalize](https://github.com/Meteor-Community-Packages/denormalize) package which uses this logic to cache values.

If you have any ideas here, I'll be glad to try something to fix this.

**Reply by msavin** [reply]
Overall it looks really solid... I'll be trying it soon. My top critiques would be: 

1. not clear if it runs sync / async

2. not 100% clear on the cases with async callbacks - and that leads to the big question of how can one throw an error if the transaction fails (i.e. if client is waiting for a response) - and would be nice to have some examples of what kind of problems one can expect

3. it would be nice if you can return some data from the function but maybe example below is not realistic

```javascript
const { invoiceId, lineItemsId, paymentsId } = runInTransaction(() => {
        const invoiceId = Invoice.insert(invoice);
        const lineItemsId = LineItems.insert(lineItems);

        const payment = createPaymentDoc(); 
        const paymentsId = Payments.insert(payment);

        return { invoiceId, lineItemsId, paymentsId };
});
```
4. purely cosmetic, but I'd change runInTransaction to just runTransaction or Transaction

**Reply by bhunjadi** [reply]
Thanks for the input! I'll update the docs accordingly. 

1. It is sync currently, meaning fibers code will run ok. await / async will not, but I have some ideas how to solve this.
2. Got it
3. You can return, the example you gave should work
4. I'll consider it after everything above is done

**Reply by bhunjadi** [reply]
I updated readme and I think I addressed all of your questions and comments in points 1-3.
Not sure I see a clean solution that fits all use cases for async callback issues I mentioned in new version of readme.

I plan to check how to support async/await in combination with fibers since this is the direction Meteor is heading now. I think it might be possible.

**Reply by msavin** [reply]
Just took a look - I notice you are throwing an error manually in the description - but how does the runInTransaction call throw an error if db server is unavailable, etc?

**Reply by bhunjadi** [reply]
I deliberately used `throw new Error()` since this is the way all errors are thrown inside the fibers. `Invoice.insert()` Meteor will throw an error the same way as I did.

As for how, `runInTransaction` catches this, aborts the transaction and throws again so you can catch it.

Of course, if some of the errors are from async functions, then `runInTransaction` won't catch them.
But you won't also with this:
```
try {
  Invoice.insert({})
} catch(e) {
}
```

**Reply by bhunjadi** [reply]
I did some more investigation regarding support for promises and async / await, but it appears that the method I was thinking on going is not possible as long as Meteor uses fibers.
Details are here: https://github.com/bhunjadi/meteor-mongo-transactions/issues/3


Besides this, I made progress regarding async callbacks and managed to handle all the test cases that were failing up til now, more details in [this PR](https://github.com/bhunjadi/meteor-mongo-transactions/pull/4)
All tests are now passing.

**Reply by msavin** [reply]
@bhunjadi I believe regular MongoDB operations in Meteor support manual callbacks for errors - so for example: 

```js
Invoices.insert({
	...data
}, (e,r) => {
	// callback
}
```

It would be nice if a callback can be passed in similarly, and then to have some controls to retry the transaction or just ensure it is aborted, etc.

**Reply by bhunjadi** [reply]
This is what I managed to solve in the PR, looks at the tests [here](https://github.com/bhunjadi/meteor-mongo-transactions/blob/b1d7b6af8b0a3e96b2ded0ac660e3da41fd47852/tests/server/server.tests.ts#L216)

They were failing previously since `runInTransaction` did not know how to wait for these callbacks.

**Reply by Torgen** [reply]
It looks like you haven't published your package, and someone else who published from a fork isn't up to date. Are you planning to publish, or should I `git submodule` it into my packages directory?

**Reply by bhunjadi** [reply]
Yeah, I'm using it as a submodule right now. No plan to publish it for now.

**Reply by Torgen** [reply]
I forked your repo and added trivial client-side stubs so I could use transactions on the server side without losing simulation on the client side or having to duplicate the method implementation. I don't think they're generally useful, since there are corner cases I never run into, but that are possible in general. For example, if halfway through the transaction you discover that some invariant is violated, you can throw an exception. In simulation, the changes won't be rolled back immediately, but will be discarded when the server-side method call finishes. However, there's an option to Meteor.apply to abort the call if the simulation throws an error; in that case the partially applied simulation will be kept around even though the server call will never be made. I think this leaks memory since it's keeping track of a method call that will never be made, but since the functionality exists in the framework, it needs to be supported. This is also an argument for Meteor to provide an interface to transactions natively: It already knows how to roll back a change on the client, since it can roll back the simulated method results, so the framework can reuse that behavior for transactions in a way that a package can't.

**Comment by jkasevits** [comment]
I might be misunderstanding something here, but the question I have is why not make use of the official node mongo driver ([rawCollection](https://docs.meteor.com/api/collections.html#Mongo-Collection-rawCollection)) to handle a case like this?

The fact that Meteor duplicates some of the official driver's API provides some benefits, but also comes at a cost. The benefits being that it works with minimongo, so enables optimistic updates, and operations are synchronous via Fibers, like most Meteor code. But the cost is rather heavy. Meteor needs to play catch-up with the offical node mongo driver and it will never be as feature complete as the official driver.

Mongo transactions have been available for quite some time already for Meteor users using rawCollection. Yes, it requires the use of async code. But it is already clear that Meteor will need to get rid of Fibers soon. So it seems to me that it does not make sense to add any new Fibers based code anymore. And if async will be required anyway, using rawCollection for transactions seems like an easy step.

As a more general note, I'd like to see Meteor promote the use of rawCollection a bit more, for example in the docs. Relying on the offical driver should be a very stable and tested way to interact with the db in Meteor, as Meteor apps are just node app at the end of the day. I don't think it should be regarded as using some obscure internal API that might break at any time.

**Reply by bhunjadi** [reply]
If I'd start a new project with Meteor or have a small codebase in an existing project, I'd go with async code as you said. Right now I don't see why I'd even use Meteor's `rawCollection()` instead of `mongodb` driver directly since it looks to me that Meteor adds nothing on top of `rawCollection()`, except another fiber to make it synchronous.

Using `rawCollection()` means that you can no longer rely on packages like [meteor-collection2](https://github.com/Meteor-Community-Packages/meteor-collection2) which is useful for validation. Again, starting from scratch you might have another option, but if you have a  large codebase, this might be an issue.

Also, if you're using some other package that writes to the DB in you transactions you have to pass `session` there and, of course, package must support it. (I think this might be solved with `async_hooks`, though)

Right now I think we don't know what will post-fiber Meteor look like, there's still [discussion](https://github.com/meteor/meteor/discussions/11505) about this. Therefore, Meteor doesn't have a clear path to no fiber future, and if you need transactions today you might go your own way with async code or try to figure it out how to make it work with fibers as long as there is no definite answer how will they migrate to async/await.

I'm in the latter group and just wanted to give my perspective here on why I'm still inclined to use fibers.

**Reply by make-github-pseudonymous-again** [reply]
> Right now I don't see why I'd even use Meteor's `rawCollection()` instead of `mongodb` driver directly since it looks to me that Meteor adds nothing on top of `rawCollection()`, except another fiber to make it synchronous.

Client minimongo simulation (actually client/server isomorphism in general): See https://github.com/meteor/meteor/pull/11605#issuecomment-957754139

**Reply by bhunjadi** [reply]
Yeah, but I'm taking here about using `rawCollection()` vs `mongodb` driver directly. I checked what does that do on a client and you get `Can only call rawCollection on server collection`.

**Reply by make-github-pseudonymous-again** [reply]
OK complete brain-fart on my side. I humbly ask for forgiveness.

You got to have a way to refer to the-mongodb-driver-collection-object-that-corresponds-to-the-meteor-collection-object though. Although you can probably cache the mapping instead of littering your codebase with `.rawCollection()` method calls?

**Reply by davidsavoie1** [reply]
I succesfully used `Coll.rawCollection()` and `session.withTransaction` from the NodeJS driver. Everything seemed to work very well... until using it with a remote MongoDB database.

With a local instance of MongoDB, the transaction writes to the database correctly and the changes are caught fast by the publications. However, with a remote database, a transaction could take _up to 7 seconds_ before being reflected in the UI. It seems to the user that nothing is happpening at all for this entire duration.

An answer from @znewsham [explains very well why this is happening](https://forums.meteor.com/t/transactions-are-too-slow-very-slow/49916/10?u=davidsavoie1). When using the Meteor collection methods, the changes are propagated fast to the connected clients, bypassing the database oplog notification. When using the native collection, Meteor doesn't know about the changes being made and **will only react to the oplog**. 

This is **fine on a single member replica set** (local MongoDB), but will _kill reactivity_ on a remote, production grade multi-cluster database.

So... The Meteor "magic" seems to come from the wrapping of collections. Please correct me if I understand it wrong.

I would be most grateful to get a simple solution to this problem.

**Reply by jamauro** [reply]
@davidsavoie1 would you mind embedding a screenshot of the znewsham answer you're referring to? I can't access that Microsoft teams link. Thanks!

**Reply by Torgen** [reply]
@davidsavoie1 See my answer in another thread in this discussion: https://github.com/meteor/meteor/discussions/11737#discussioncomment-3831357

**Reply by davidsavoie1** [reply]
Oups! Sorry @jamauro, I hadn't notice the erroneous link. I fixed it...

**Comment by a4xrbj1** [comment]
Are there any updates on this feature request or more clarity in terms of fiber and MongoDb?

**Comment by jamauro** [comment]
I just published a new package [jam:mongo-transactions](https://github.com/jamauro/mongo-transactions) that was inspired by this discussion and builds on some of the ideas shared here. Here are some of the benefits:

* Write with the standard Meteor collection methods you're accustomed to. You don't need to worry about using `rawCollection()`.
* Because it's a low-level solution, ID generation works as expected.
* You don't need to worry about passing `session` around. This package takes care of that for you.
* Works out-of-the-box with other packages that automatically validate on DB writes, like `jam:easy-schema` and `aldeed:collection2`.
* One simple API to use. Mongo has made things complicated with two APIs for Transactions, the Callback API and the Core API. This package defaults to using the Callback API as recommended by Mongo, but allows you to use the Core API by passing in `autoRetry: false`.
* Can be used isomorphically.
* Compatible with Meteor 2.8.1 and up, including support for Meteor 3.0+

Here's a quick example. As said above, you can surf the async vibe:

```js
import { Mongo } from 'meteor/mongo';

async function purchase(purchaseData) {
  try {
    const { invoiceId } = await Mongo.withTransaction(async () => {
      const invoiceId = await Invoices.insertAsync(purchaseData);
      const changeQuantity = await Items.updateAsync(purchaseData.itemId, { $set: {...} });
      return { invoiceId, changeQuantity } // you can return whatever you'd like
    });
    return invoiceId;
  } catch (error) {
    // something went wrong with the transaction and it could not be automatically retried
    // handle the error as you see fit
  }
}
```

If you end up taking it for a spin or have ideas on how to make it better, let me know!

**Reply by StorytellerCZ** [reply]
This should be in core 😉

**Reply by Torgen** [reply]
Do you have any issues with the results of transactions not being immediately visible to the client who called the method? Since `Meteor.refresh` is called before the transaction is committed (because it's part of Meteor's wrappers around `RawCollection`), I don't know whether the subscriptions will look for updates at that time, not see them because the transaction isn't committed yet, then not check again when the transaction commits and the changes are visible.

**Reply by jamauro** [reply]
@Torgen No, I didn’t see any issues there in my testing. The subscription will pick up the changes to the publication after the transaction commits.

**Reply by make-github-pseudonymous-again** [reply]
@jamauro This implementation is using `Meteor.EnvironmentVariable` to keep track of the `session` object. Isn't this going to fail for concurrent async calls (no fibers)?

**Reply by jamauro** [reply]
@make-github-pseudonymous-again the attempts to test concurrency are [here](https://github.com/jamauro/mongo-transactions/blob/56719f31b1e5efcec045d505adf8554db19653c9/mongo-transactions-tests.js#L718) and there's another below it. They are passing on `Meteor 3.0-beta.0` and I can see that a unique `session` object is being created for each transaction and carried through to the operations within it. Let me know if you were thinking of another way to test it.

**Reply by jamauro** [reply]
> This should be in core 😉

Agree, ultimately I'd like to see this in Meteor core. Would be great to hear what the Meteor core team thinks. @fredmaiaarantes @denihs @Grubba27 

More generally, I think we shouldn't need to reach for `rawCollection`. I posted a [discussion](https://github.com/meteor/meteor/discussions/12779) a while back to that end. Hopefully, after 3.0 is out, these gaps can be addressed.

**Reply by make-github-pseudonymous-again** [reply]
@jamauro Does it work because it uses two different connections then? I guess this is the only way to achieve parallelism anyway. I must admit I am not really sure I understand what these "environment variables" are tied to in fiber-less Meteor.
